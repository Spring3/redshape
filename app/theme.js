export default {
  bg: '#FFFFFF',
  bgLight: '#F9F9F9',
  mainLight: '#FF7F87',
  main: '#FF7079',
  mainDark: '#EF6069',
  minorText: '#909090',
  normalText: '#3F3844',
  hoverText: '#FFFFFF',
  red: '#FF634D',
  yellow: '#FFDA77',
  green: '#6CCA51',
  transitionTime: '.2s'
};
