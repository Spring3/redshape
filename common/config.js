const storage = require('./storage');
const env = require('./env');

let config = { ...env };

if (storage.has('settings')) {
  config = { ...config, ...storage.get('settings') };
}

module.exports = config;
