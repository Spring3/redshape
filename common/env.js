const { remote } = require('electron');
const _ = require('lodash');

let config;

// main process
const { env } = (process || remote.process);

if (!config) {
  config = _.pick(env, 'PORT', 'ENCRYPTION_KEY', 'NODE_ENV');
}

module.exports = config;
