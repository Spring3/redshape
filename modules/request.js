const axios = require('axios');
const storage = require('./storage');

let instance;

const TWENTY_SECONDS = 20000;

const pendingRequests = {};

const makeCancellable = (headers, requestId) => {
  if (!requestId) {
    return headers;
  }

  if (pendingRequests[requestId]) {
    pendingRequests[requestId].cancel();
  }
  const source = axios.CancelToken.source();
  const newHeaders = {
    ...headers,
    cancelToken: source.token
  };
  pendingRequests[requestId] = source;
  return newHeaders;
};

const initialize = (redmineEndpoint, token) => {
  if (!redmineEndpoint || !token) {
    return;
  }
  const axiosConfig = {
    baseURL: redmineEndpoint,
    timeout: TWENTY_SECONDS,
    headers: {
      'X-Redmine-API-Key': token
    }
  };
  instance = axios.create(axiosConfig);
};

const getInstance = () => instance;

const reset = () => { instance = undefined; };

if (storage.has('user')) {
  const { api_key, redmineEndpoint } = storage.get('user');
  if (api_key && redmineEndpoint) {
    initialize(redmineEndpoint, api_key);
  }
}

const handleReject = (error) => {
  // if this request was not cancelled
  if (!axios.isCancel(error)) {
    let errorMessage = 'Error';
    if (error.status) {
      errorMessage = `${errorMessage} ${error.status}`;
    }
    errorMessage = `${errorMessage} (${error.statusText || error.message})`;

    return Promise.reject(new Error(errorMessage));
  }
  return undefined;
};

const authorizedRequest = (config) => {
  if (config.id) {
    Object.assign(
      config,
      { headers: makeCancellable(config.headers || {}, config.id) }
    );
  }
  if (!instance) {
    return Promise.reject(new Error('401 - Unauthorized'));
  }

  Object.assign(
    config,
    {
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, PATCH, OPTIONS',
        'Access-Control-Allow-Headers': 'X-Requested-With, content-type, Authorization',
        'X-Redmine-API-Key': instance.defaults.headers['X-Redmine-API-Key'],
        ...config.headers
      }
    }
  );

  return getInstance().request(config)
    .then((res) => {
      delete pendingRequests[config.id];
      return ({ data: res.data });
    })
    .catch((error) => {
      delete pendingRequests[config.id];
      return handleReject(error);
    });
};

const request = (config) => {
  Object.assign(config, {
    timeout: TWENTY_SECONDS,
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, PATCH, OPTIONS',
      'Access-Control-Allow-Headers': 'X-Requested-With, content-type, Authorization',
      ...config.headers
    }
  });
  return axios.request(config).catch(handleReject);
};

module.exports = {
  default: axios,
  makeCancellable,
  initialize,
  getInstance,
  pendingRequests,
  reset,
  authorizedRequest,
  request
};
