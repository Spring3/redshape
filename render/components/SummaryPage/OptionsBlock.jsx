import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled from 'styled-components';

import actions from '../../actions';
import { Input, Label } from '../Input';

const MarginedDiv = styled.div`
  margin-top: 10px;
`;

class OptionsBlock extends Component {
  toggleClosedIssuesDisplay = () => {
    const { settingsShowClosedIssues, showClosedIssues } = this.props;
    settingsShowClosedIssues(!showClosedIssues);
  }

  toggleUseColors = () => {
    const { settingsUseColors, useColors } = this.props;
    settingsUseColors(!useColors);
  }

  render() {
    const { showClosedIssues, useColors } = this.props;
    return (
      <div>
        <Label htmlFor="queryOptions" label="Options">
          <div id="queryOptions">
            <label>
              <Input
                type="checkbox"
                checked={showClosedIssues}
                onChange={this.toggleClosedIssuesDisplay}
              />
              <span>Include Closed</span>
            </label>
          </div>
        </Label>
        <MarginedDiv>
          <label>
            <Input
              type="checkbox"
              checked={useColors}
              onChange={this.toggleUseColors}
            />
            <span>Use Colors</span>
          </label>
        </MarginedDiv>
      </div>
    );
  }
}

OptionsBlock.propTypes = {
  showClosedIssues: PropTypes.bool.isRequired,
  useColors: PropTypes.bool.isRequired,
  settingsShowClosedIssues: PropTypes.func.isRequired,
  settingsUseColors: PropTypes.func.isRequired
};

const mapStateToProps = (state) => ({
  showClosedIssues: state.settings.showClosedIssues,
  useColors: state.settings.useColors
});

const mapDispatchToProps = (dispatch) => ({
  settingsShowClosedIssues: (value) => dispatch(actions.settings.setShowClosedIssues(value)),
  settingsUseColors: (value) => dispatch(actions.settings.setUseColors(value))
});

export default connect(mapStateToProps, mapDispatchToProps)(OptionsBlock);
