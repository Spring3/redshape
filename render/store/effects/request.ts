import { ipcRenderer, remote } from 'electron';
import { Response } from '../../../types';

const crypto = remote.require('crypto');

type Config = {
  endpoint: string;
  token?: string;
}

type QueryConfig = {
  payload: Record<string, any>;
  config?: Config;
}

const query = (config: QueryConfig): Promise<Response> => {
  const id = crypto.randomBytes(10).toString('hex');

  return new Promise((resolve) => {
    ipcRenderer.send('request', JSON.stringify({ ...config, id }));

    ipcRenderer.once(`response:${id}`, (event, response) => {
      console.log('response:', response);
      resolve(response);
    });
  });
};

export {
  query,
};

export type {
  QueryConfig
};
