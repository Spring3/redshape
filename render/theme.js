export default {
  bg: '#FFFFFF',
  bgDark: '#F9F9F9',
  bgDarker: '#E9E9E9',
  bgDisabled: 'hsl(0,0%,95%)',
  mainLight: '#FF7F87',
  main: '#FF7079',
  mainDark: '#EF6069',
  minorText: '#A0A0A0',
  normalText: '#3F3844',
  hoverText: '#FFFFFF',
  red: '#FF634D',
  yellow: '#FFDA77',
  green: '#6CCA51',
  shadow: '#D0D0D0',
  transitionTime: '.2s'
};
