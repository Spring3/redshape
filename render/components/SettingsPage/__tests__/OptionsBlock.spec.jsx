import React from 'react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import MockAdapter from 'axios-mock-adapter';

import {
  SETTINGS_SHOW_ADVANCED_TIMER_CONTROLS,
  SETTINGS_UI_STYLE,
  SETTINGS_IDLE_BEHAVIOR,
  SETTINGS_PROGRESS_SLIDER,
  SETTINGS_IDLE_TIME_DISCARD, SETTINGS_ISSUE_ALWAYS_EDITABLE, SETTINGS_TIMER_CHECKPOINT,
} from '../../../actions/settings.actions';
import OptionsBlock from '../OptionsBlock';
import { getInstance, reset, initialize } from '../../../../common/request';
import theme from '../../../theme';

const redmineEndpoint = 'redmint.test.test';
const token = 'multipass';
const mockStore = configureStore([thunk]);

let axiosMock;

describe('SummaryPage => OptionsBlock component', () => {
  beforeAll(() => {
    initialize(redmineEndpoint, token);
    axiosMock = new MockAdapter(getInstance());
    axiosMock.onAny('.json').reply(() => Promise.resolve([200]));
  });

  afterEach(() => {
    axiosMock.reset();
  });

  afterAll(() => {
    axiosMock.restore();
    reset();
  });

  it('should match the snapshot', () => {
    const state = {
      settings: {
        uiStyle: 'colors',
        showClosedIssues: true
      },
      user: {
        id: 1
      }
    };
    const store = mockStore(state);
    const tree = renderer.create(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <OptionsBlock />
        </ThemeProvider>
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should set use of advanced timer controls when select option is chosen', () => {
    const state = {
      settings: {
        showAdvancedTimerControls: false,
        showClosedIssues: true
      },
      user: {
        id: 1,
        redmineEndpoint: 'https://redmine.redmine'
      }
    };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <OptionsBlock />
        </ThemeProvider>
      </Provider>
    );

    wrapper.find('input[type="checkbox"]').at(0).simulate('change', { target: { checked: true } });
    expect(store.getActions().length).toBe(1);
    expect(store.getActions()[0]).toEqual({
      type: SETTINGS_SHOW_ADVANCED_TIMER_CONTROLS,
      data: {
        userId: state.user.id,
        redmineEndpoint: state.user.redmineEndpoint,
        showAdvancedTimerControls: true
      }
    });
  });

  it('should set use of issue always editable when select option is chosen', () => {
    const state = {
      settings: {
        showAdvancedTimerControls: false,
        showClosedIssues: true,
        isIssueAlwaysEditable: false,
      },
      user: {
        id: 1,
        redmineEndpoint: 'https://redmine.redmine'
      }
    };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <OptionsBlock />
        </ThemeProvider>
      </Provider>
    );

    wrapper.find('input[type="checkbox"]').at(1).simulate('change', { target: { checked: true } });
    expect(store.getActions().length).toBe(1);
    expect(store.getActions()[0]).toEqual({
      type: SETTINGS_ISSUE_ALWAYS_EDITABLE,
      data: {
        userId: state.user.id,
        redmineEndpoint: state.user.redmineEndpoint,
        isIssueAlwaysEditable: true
      }
    });
  });

  it('should set use of idle behavior when select option is chosen', () => {
    const state = {
      settings: {
        showAdvancedTimerControls: false,
        showClosedIssues: true,
        uiStyle: 'default',
        idleBehavior: 'none',
        idleTimeDiscard: false,
      },
      user: {
        id: 1,
        redmineEndpoint: 'https://redmine.redmine'
      }
    };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <OptionsBlock />
        </ThemeProvider>
      </Provider>
    );

    const selects = wrapper.find('Select');
    selects.at(0).instance().selectOption({ value: '15m' });
    expect(store.getActions().length).toBe(1);
    expect(store.getActions()[0]).toEqual({
      type: SETTINGS_IDLE_BEHAVIOR,
      data: {
        userId: state.user.id,
        redmineEndpoint: state.user.redmineEndpoint,
        idleBehavior: '15m'
      }
    });

    wrapper.find('input[type="checkbox"]').at(2).simulate('change', { target: { checked: true } });
    expect(store.getActions().length).toBe(2);
    expect(store.getActions()[1]).toEqual({
      type: SETTINGS_IDLE_TIME_DISCARD,
      data: {
        userId: state.user.id,
        redmineEndpoint: state.user.redmineEndpoint,
        idleTimeDiscard: true
      }
    });

    selects.at(0).instance().selectOption({ value: 'none' });
    expect(store.getActions().length).toBe(3);
    expect(store.getActions()[2]).toEqual({
      type: SETTINGS_IDLE_BEHAVIOR,
      data: {
        userId: state.user.id,
        redmineEndpoint: state.user.redmineEndpoint,
        idleBehavior: 'none'
      }
    });

    expect(wrapper.find('input[type="checkbox"]').at(2).props().checked).toBe(false);

    selects.at(0).instance().selectOption({ label: 'Pause if idle for 5 minutes', value: '5m' });
    expect(store.getActions().length).toBe(4);
    expect(store.getActions()[3]).toEqual({
      type: SETTINGS_IDLE_BEHAVIOR,
      data: {
        userId: state.user.id,
        redmineEndpoint: state.user.redmineEndpoint,
        idleBehavior: '5m'
      }
    });
  });

  it('should set use of timer checkpoint when select option is chosen', () => {
    const state = {
      settings: {
        showAdvancedTimerControls: false,
        showClosedIssues: true,
        uiStyle: 'default',
        idleBehavior: 'none',
        progressSlider: '10%',
        timerCheckpoint: 'none',
      },
      user: {
        id: 1,
        redmineEndpoint: 'https://redmine.redmine'
      }
    };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <OptionsBlock />
        </ThemeProvider>
      </Provider>
    );

    const selects = wrapper.find('Select');
    selects.at(1).instance().selectOption({ value: '10m' });
    expect(store.getActions().length).toBe(1);
    expect(store.getActions()[0]).toEqual({
      type: SETTINGS_TIMER_CHECKPOINT,
      data: {
        userId: state.user.id,
        redmineEndpoint: state.user.redmineEndpoint,
        timerCheckpoint: '10m'
      }
    });
  });

  it('should set use of progress slider when select option is chosen', () => {
    const state = {
      settings: {
        showAdvancedTimerControls: false,
        showClosedIssues: true,
        uiStyle: 'default',
        idleBehavior: 'none',
        progressSlider: '10%',
      },
      user: {
        id: 1,
        redmineEndpoint: 'https://redmine.redmine'
      }
    };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <OptionsBlock />
        </ThemeProvider>
      </Provider>
    );

    const selects = wrapper.find('Select');
    selects.at(2).instance().selectOption({ value: '1%' });
    expect(store.getActions().length).toBe(1);
    expect(store.getActions()[0]).toEqual({
      type: SETTINGS_PROGRESS_SLIDER,
      data: {
        userId: state.user.id,
        redmineEndpoint: state.user.redmineEndpoint,
        progressSlider: '1%'
      }
    });
  });

  it('should set use of ui style when select option is chosen', () => {
    const state = {
      settings: {
        showAdvancedTimerControls: false,
        showClosedIssues: true,
        uiStyle: 'default',
      },
      user: {
        id: 1,
        redmineEndpoint: 'https://redmine.redmine'
      }
    };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <OptionsBlock />
        </ThemeProvider>
      </Provider>
    );

    const selects = wrapper.find('Select');
    selects.at(3).instance().selectOption({ label: 'Default with colors', value: 'colors' });
    expect(store.getActions().length).toBe(1);
    expect(store.getActions()[0]).toEqual({
      type: SETTINGS_UI_STYLE,
      data: {
        userId: state.user.id,
        redmineEndpoint: state.user.redmineEndpoint,
        uiStyle: 'colors'
      }
    });
  });

});
