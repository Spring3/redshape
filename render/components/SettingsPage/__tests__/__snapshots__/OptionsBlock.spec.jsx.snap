// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryPage => OptionsBlock component should match the snapshot 1`] = `
.c10 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clippath: inset(50%);
  clippath: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c11 {
  display: inline-block;
  width: 16px;
  height: 16px;
  background: #F9F9F9 transition:background .2s;
  border: 2px solid transparent;
  border-radius: 3px;
  cursor: pointer;
  background: white;
  border-color: #FF7079;
}

.c11 svg {
  visibility: hidden;
  position: relative;
  vertical-align: middle;
  bottom: 3px;
}

.c9:not(:disabled):hover + .c11,
.c9:focus + .c11 {
  box-shadow: 0px 0px 5px 1px #FF7F87;
}

.c9:disabled + .c11 {
  background: #F9F9F9;
  border-color: lightgrey;
}

.c8 {
  display: inline-block;
  vertical-align: middle;
}

.c8 + span {
  margin-left: 10px;
  vertical-align: middle;
}

.c1 h4 {
  margin-bottom: 10px;
  color: #A0A0A0;
}

.c2 {
  color: #A0A0A0;
}

.c7 {
  position: absolute;
  display: none;
  border-radius: 3px;
  width: auto;
  padding: 5px;
  background: #FFFFFF;
  color: #EF6069;
  text-align: center;
  font-size: 12px;
  white-space: nowrap;
  box-shadow: 0px 2px 7px #A0A0A0;
  z-index: 99;
  top: -100%;
  left: 30px;
}

.c7::after {
  content: ' ';
  position: absolute;
  width: auto;
  border: 2px solid black;
  border-width: 5px;
  border-color: #FFFFFF transparent transparent transparent;
}

.c7::after {
  top: calc(50% - 5px);
  left: -10px;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.c12 {
  position: absolute;
  display: none;
  border-radius: 3px;
  width: auto;
  padding: 5px;
  background: #FFFFFF;
  color: #EF6069;
  text-align: center;
  font-size: 12px;
  white-space: nowrap;
  box-shadow: 0px 2px 7px #A0A0A0;
  z-index: 99;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  top: -50px;
  left: 50%;
}

.c12::after {
  content: ' ';
  position: absolute;
  width: auto;
  border: 2px solid black;
  border-width: 5px;
  border-color: #FFFFFF transparent transparent transparent;
}

.c12::after {
  top: 100%;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c4 {
  display: inline-block;
  position: relative;
}

.c4:hover .c6 {
  display: block;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  padding-bottom: 1px;
}

.c3 {
  margin-left: 0.2rem;
  color: #A0A0A0;
}

.c0 {
  display: grid;
  grid-template-columns: minmax(auto,400px) minmax(auto,500px);
}

@media (max-width:1200px) {
  .c0 {
    grid-template-columns: 1fr 2fr;
  }
}

@media (max-width:950px) {
  .c0 {
    grid-template-columns: 1fr;
  }
}

<div
  className="c0"
>
  <div
    className="form-group undefined c1"
  >
    <h4
      className="c2"
      htmlFor="queryOptions"
    >
      Options
      <span
        className="c3"
      >
        <div
          className="c4"
        >
          <svg
            className="mdi-icon c5"
            fill="currentColor"
            height={14}
            viewBox="0 0 24 24"
            width={14}
          >
            <path
              d="M15.07,11.25L14.17,12.17C13.45,12.89 13,13.5 13,15H11V14.5C11,13.39 11.45,12.39 12.17,11.67L13.41,10.41C13.78,10.05 14,9.55 14,9C14,7.89 13.1,7 12,7A2,2 0 0,0 10,9H8A4,4 0 0,1 12,5A4,4 0 0,1 16,9C16,9.88 15.64,10.67 15.07,11.25M13,19H11V17H13M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12C22,6.47 17.5,2 12,2Z"
            />
          </svg>
          <p
            className="tooltip c6 c7"
          >
            - Advanced timer controls: timer and comment are modifiable at runtime.
          </p>
        </div>
      </span>
    </h4>
    <div
      id="queryOptions"
    >
      <label>
        <div
          className="c8"
        >
          <input
            checked={false}
            className="c9 c10"
            disabled={false}
            onChange={[Function]}
            type="checkbox"
          />
          <div
            checked={false}
            className="c11"
          >
            <svg
              className="mdi-icon "
              fill="white"
              height="18"
              viewBox="0 0 24 24"
              width="18"
            >
              <path
                d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"
              />
            </svg>
          </div>
        </div>
        <span>
          Use advanced timer controls
        </span>
      </label>
    </div>
  </div>
  <div
    className="form-group undefined c1"
  >
    <h4
      className="c2"
      htmlFor="idleBehavior"
    >
      Timer behavior
      <span
        className="c3"
      >
        <div
          className="c4"
        >
          <svg
            className="mdi-icon c5"
            fill="currentColor"
            height={14}
            viewBox="0 0 24 24"
            width={14}
          >
            <path
              d="M15.07,11.25L14.17,12.17C13.45,12.89 13,13.5 13,15H11V14.5C11,13.39 11.45,12.39 12.17,11.67L13.41,10.41C13.78,10.05 14,9.55 14,9C14,7.89 13.1,7 12,7A2,2 0 0,0 10,9H8A4,4 0 0,1 12,5A4,4 0 0,1 16,9C16,9.88 15.64,10.67 15.07,11.25M13,19H11V17H13M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12C22,6.47 17.5,2 12,2Z"
            />
          </svg>
          <p
            className="tooltip c6 c12"
          >
            Detect if the system is idle to pause the timer.
          </p>
        </div>
      </span>
    </h4>
    <div
      className="c13"
    >
      <div
        className="css-3q53pl"
        onKeyDown={[Function]}
      >
        <div
          className="css-1rcfx4g"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className="css-1hwfws3"
          >
            <div
              className="css-1492t68"
            >
              Select...
            </div>
            <div
              className="css-1g6gooi"
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-autocomplete="list"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  disabled={false}
                  id="react-select-2-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  spellCheck="false"
                  style={
                    Object {
                      "background": 0,
                      "border": 0,
                      "boxSizing": "content-box",
                      "color": "inherit",
                      "fontSize": "inherit",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "1px",
                    }
                  }
                  tabIndex="0"
                  type="text"
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className="css-1wy0on6"
          >
            <span
              className="css-d8oujb"
            />
            <div
              aria-hidden="true"
              className="css-1ep9fjw"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-19bqh2r"
                focusable="false"
                height={20}
                viewBox="0 0 20 20"
                width={20}
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="idleBehavior"
          type="hidden"
          value=""
        />
      </div>
      <label
        className="c14"
      >
        <div
          className="c8"
        >
          <input
            checked={false}
            className="c9 c10"
            disabled={false}
            onChange={[Function]}
            type="checkbox"
          />
          <div
            checked={false}
            className="c11"
          >
            <svg
              className="mdi-icon "
              fill="white"
              height="18"
              viewBox="0 0 24 24"
              width="18"
            >
              <path
                d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"
              />
            </svg>
          </div>
        </div>
        <span>
          Auto discard idle time from timer
        </span>
      </label>
    </div>
  </div>
  <div
    className="c13"
  >
    <div
      className="form-group undefined c1"
    >
      <h4
        className="c2"
        htmlFor="progress"
      >
        Progress values
        <span
          className="c3"
        >
          <div
            className="c4"
          >
            <svg
              className="mdi-icon c5"
              fill="currentColor"
              height={14}
              viewBox="0 0 24 24"
              width={14}
            >
              <path
                d="M15.07,11.25L14.17,12.17C13.45,12.89 13,13.5 13,15H11V14.5C11,13.39 11.45,12.39 12.17,11.67L13.41,10.41C13.78,10.05 14,9.55 14,9C14,7.89 13.1,7 12,7A2,2 0 0,0 10,9H8A4,4 0 0,1 12,5A4,4 0 0,1 16,9C16,9.88 15.64,10.67 15.07,11.25M13,19H11V17H13M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12C22,6.47 17.5,2 12,2Z"
              />
            </svg>
            <p
              className="tooltip c6 c12"
            >
              Use 10% unless you have server-side support.
            </p>
          </div>
        </span>
      </h4>
      <div
        className="css-3q53pl"
        onKeyDown={[Function]}
      >
        <div
          className="css-1rcfx4g"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className="css-1hwfws3"
          >
            <div
              className="css-1492t68"
            >
              Select...
            </div>
            <div
              className="css-1g6gooi"
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-autocomplete="list"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  disabled={false}
                  id="react-select-3-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  spellCheck="false"
                  style={
                    Object {
                      "background": 0,
                      "border": 0,
                      "boxSizing": "content-box",
                      "color": "inherit",
                      "fontSize": "inherit",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "1px",
                    }
                  }
                  tabIndex="0"
                  type="text"
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className="css-1wy0on6"
          >
            <span
              className="css-d8oujb"
            />
            <div
              aria-hidden="true"
              className="css-1ep9fjw"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-19bqh2r"
                focusable="false"
                height={20}
                viewBox="0 0 20 20"
                width={20}
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="progress"
          type="hidden"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="c13"
  >
    <div
      className="form-group undefined c1"
    >
      <h4
        className="c2"
        htmlFor="uiStyle"
      >
        UI Style
      </h4>
      <div
        className="css-3q53pl"
        onKeyDown={[Function]}
      >
        <div
          className="css-1rcfx4g"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className="css-1hwfws3"
          >
            <div
              className="css-xp4uvy"
            >
              Default with colors
            </div>
            <div
              className="css-1g6gooi"
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-autocomplete="list"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  disabled={false}
                  id="react-select-4-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  spellCheck="false"
                  style={
                    Object {
                      "background": 0,
                      "border": 0,
                      "boxSizing": "content-box",
                      "color": "inherit",
                      "fontSize": "inherit",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "1px",
                    }
                  }
                  tabIndex="0"
                  type="text"
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className="css-1wy0on6"
          >
            <span
              className="css-d8oujb"
            />
            <div
              aria-hidden="true"
              className="css-1ep9fjw"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-19bqh2r"
                focusable="false"
                height={20}
                viewBox="0 0 20 20"
                width={20}
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="uiStyle"
          type="hidden"
          value="colors"
        />
      </div>
    </div>
  </div>
</div>
`;
